mgr/dashboard: Pool management

Devp00l opened the pull request:
This PR has a backend change in it, so please before testing the frontend reload the backend. If you are not reloading the backend the tasks will missmatch, this way you will receive `Updated 'undefined'` or no task at all.

You can create a pool through the form.

You can specify the following:
* Name - can't be duplicated
* Type - replicated or erasure
* Crush rule set - Validates if you can use it
* Replica size - depends on your selected rule and the ammount of osds
* Erasure code profile
* PGs - will be recalculated on form changes (type, replica size,
    erasure profile, crush rule)
* EC overwrites flag
* Compression - Algorithm / Min/max blob size / mode / ratio
* Application metadata - Predefined and custom applications

Signed-off-by: Stephan Müller <smueller@suse.com>

# Erasure code pool creation
![ec-pool-creation](https://user-images.githubusercontent.com/16167865/43504606-62848d84-9564-11e8-9b0e-93d1b4dd8620.gif)

# Replicated pool creation
![replicated-pool-creation](https://user-images.githubusercontent.com/16167865/43504608-62e565b4-9564-11e8-93ce-9071744d0e99.gif)

# Edit replicated pool
![replicated-pool-edit](https://user-images.githubusercontent.com/16167865/43504609-63013848-9564-11e8-8b4f-978e61b7a46f.gif)

# Pool deletion
![pool-delete](https://user-images.githubusercontent.com/16167865/43504607-62c9304c-9564-11e8-8541-3af3fa3f7470.gif)

Devp00l commented:
This PR is still WIP, so the deletion and editing part will be added as soon as I'm confident that all options are created as they should. BTW the deletion and editing part is the smallest part. Also there are still files from the erasure code management that will be torn out of this branch soon.

djgalloway commented:
jenkins test make check

Devp00l commented:
@ricardoasmarques addressed all comments

votdev commented:
When creating a pool, selecting 'replicated' as type and press 'Create', then i get an error popup: [errno -34] pg_num 256 size 3 would mean 912 total pgs, which exceeds max 600 (mon_max_pg_per_osd 200 * num_in_osds 3). This is OK, but when i press the 'Create' button again and again, nothing happens nor any form field is marked as invalid. Would be helpful if there is any help for the user to find out what to change to get the formular happy.

votdev commented:
The UI to add applications is not intuitive. I clicked multiple times on the + button in expectation to see a list of possible types to select. I suggest to use a multi-selectable dropdown box here.

votdev commented:
When selecting a compression, then i expect that the 'Algorithm' field shows an 'required' icon. When pressing the 'Edit' or 'Add' button this field is not marked as invalid. Or is this field not required? Doesn't make sense to me then if i select mode 'aggressive' or any other mode.
If the algorithm is optional, then i expect this field to be set to a default  called 'None' or something else, but NOT 'Select a compression algorithm'.

votdev commented:
![auswahl_008](https://user-images.githubusercontent.com/1897962/43522821-c6528830-959a-11e8-9dde-777350c03690.png)
How do i process here? The 'Crush ruleset' is not selectable, the info text is really cryptic and does not enlight me how to process.
The 'Erasure code profile' dropdown isn't selectable, too. If there do no Crush rulesets exist, then i expect an error message when the dialog/form panel is shown.

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,548 @@
+<nav aria-label="breadcrumb">
+  <ol class="breadcrumb">
+    <li class="breadcrumb-item">
+      <a routerLink="/pool">Pools</a>
```
This text isn't translated

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,548 @@
+  <h1 *ngIf="!(info && ecProfiles)"
+      class="jumbotron">
+    <i class="fa fa-lg fa-pulse fa-spinner text-primary"></i>
+    Loading...
```
This text isn

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,548 @@
+                   name="name"
+                   type="text"
+                   class="form-control"
+                   placeholder="Name..."
```
i18n-placeholder is missing

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,548 @@
+              </select>
+              <span class="help-block"
+                    *ngIf="form.getValue('crushRule')">
+                Steps
```
Missing translation

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,548 @@
+                         type="text"
+                         min="0"
+                         class="form-control"
+                         ì18n-placeholder
```
s/ì18n/i18n/

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,548 @@
+                         min="0"
+                         formControlName="maxBlobSize"
+                         class="form-control"
+                         ì18n-placeholder
```
s/ì18n/i18n/

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,548 @@
+                         max="1"
+                         step="0.1"
+                         class="form-control"
+                         placeholder="Compression ratio">
```
Text is not translated

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,548 @@
+                <input type="text"
+                       [typeahead]="getAvailApps()"
+                       [typeaheadMinLength]="0"
+                       placeholder="Select or type an application name"
```
Text is not translated

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,548 @@
+                <span class="input-group-btn">
+                  <button class="btn btn-default"
+                          type="button"
+                          tooltip="Add application"
```
Text is not translated

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,548 @@
+                  <span class="input-group-btn">
+                    <button class="btn btn-default"
+                            type="button"
+                            tooltip="Remove application"
```
Text is not translated

Devp00l commented:
@votdev 
1. Compression does not require any other fields besides the mode to be modified. It's how it's implemented in Ceph.

2. I've not seen any plus icons used next to a selection or input box to populate a list. I guess you want to say that it should look like in the RGW forms or like in the user management form? The problem with how it's implemented in the user management is that I can't add custom applications there. The problem doing it with modals would be that I would need a combo box to select from or an input field, depends on how the user decides. The current way is more straightforward as you can do everything in the same field. If you go a view commits back you will see the old way to add applications like it was done in openATTIC.

3. The steps show what is done by the ruleset, but they could be hidden behind a info button? Would that suite you better?

tspmelo commented:
@Devp00l I would suggest extracting the code from user management into a component and implement the possibility to write custom options.
something like github has: 
![image](https://user-images.githubusercontent.com/399326/43639566-916c8834-9714-11e8-8ae4-47e9b25bffb4.png)

votdev commented:
Is it possible to implement the application selection like this in the user management?
![auswahl_012](https://user-images.githubusercontent.com/1897962/43714131-4d12f442-997c-11e8-8540-aa0542a70245.png)

Devp00l commented:
@votdev I'm currently doing exactly that ;)

ricardoasmarques commented:
Currently we have to press the "Enter" key to add a custom application to the list, and then select it, which will probably have some usability issues.

Is it possible to implement something like labels on github, where we can filter labels, and we have a button to create if the filter has no results?

![screenshot from 2018-08-08 10-58-02](https://user-images.githubusercontent.com/14297426/43830803-21a9f074-9afa-11e8-95ed-58ad91abc8d4.png)

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool.module.ts:
```diff
@@ -1,15 +1,30 @@
+    TabsModule,
+    PopoverModule.forRoot(),
+    SharedModule,
+    FormsModule,
```
Since we are using ReactiveForms, we don't need to import FormsModule.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table-actions/table-actions.component.html:
```diff
@@ -0,0 +1,34 @@
+      <li role="menuitem"
+          *ngIf="permission[action.permission]">
+        <a class="dropdown-item"
+           [ngClass]="{'disabled': disableSelectionAction(action.permission)}"
```
The text should be grey when disabled, but it stays black.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:
```diff
@@ -1,21 +1,56 @@
+  tableActions: CdTableAction[];
 
-  constructor(private poolService: PoolService) {
+  constructor(
```
You should subscribe to the `summaryService` and update the table each time it sends a new data.
ATM we don't see any information about running tasks related to pools.

Maybe we could extract and adapt the functions used in the `RBD List` component or simply copy them.

tspmelo replied:
Please have a look at #23555, it should help implement this.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table-actions/table-actions.component.ts:
```diff
@@ -0,0 +1,71 @@
+import { Component, Input, OnInit } from '@angular/core';
```
This should replace the current implementation of table-actions in all tables.
To do that, I would suggest creating a separate PR, since we don't want to include changes externals to Pools in this PR.
Only doing this will garante that it works in all situations and we don't need to change the code in the future.

Devp00l replied:
That was the plan ;)

tspmelo replied:
When I mentioned a separate PR, I was thinking of it being created/merged before this PR.
Are you currently working on it? or planning to?

Devp00l replied:
I can extract it to be merged before, to reduce the burden to review over 3k lines of code ;)

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/select-badges/select-badges.component.spec.ts:
```diff
@@ -11,47 +13,162 @@ describe('SelectBadgesComponent', () => {
   configureTestBed({
     declarations: [SelectBadgesComponent],
-    imports: [PopoverModule.forRoot()]
+    imports: [PopoverModule.forRoot(), FormsModule, ReactiveFormsModule]
```
There is no need to import `FormsModule`.

Devp00l commented:
I'm happy to say that I've addressed all comments :)

![pool-badges-listing](https://user-images.githubusercontent.com/16167865/44149417-51d55cf0-a09c-11e8-9095-390ca4903b12.gif)

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:
```diff
@@ -63,18 +108,71 @@ export class PoolListComponent {
-        this.pools = pools;
+      (pools: Pool[]) => {
+        this.pools = pools.map((pool) => {
+          pool.executingTasks = this.getPoolTasks(tasks, pool.pool_name);
```
Have you reloaded the backend before trying? It works for me, I don't see a reason why it should not work.

tspmelo replied:
The pool creation works.

But, if it takes a long time, it will not show up in the table while it is being created.
If you go to RBD, the moment you finished using the creation form, it will show up in the table, even while it is still being created.

try adding a sleep of 15s in both backend methods and test it.

Devp00l replied:
The creation of a pool will happen nearly instantly, the other parts will take time like settings application metadata flags and of course the creation of PGs.

Devp00l replied:
I've add the safety net, as we cannot completely predict the behavior of a huge cluster. It can be easily replaced by your PR through the added tests. There were some issues with your PR that's why I could not include it here. I'll give you the feedback soon in your PR.

Devp00l commented:
Rebased it on the current master

LenzGr commented:
One thing I observed during my testing: when creating a replicated pool, "Replicated size" (the number of replicas) defaults to "1". I may be wrong, but I think the default when creating a pool on the command line is "3". I wonder if the UI should also use this default value in order to avoid creating pools with no redundancy by accident.

Moreover, I noticed a small inconsistency in the "Placement groups" input field: when I use "cursor up/down", the number increases/decreases by the power of two, e.g. 64->128->256 etc. If I use the small up/down icons to the right of the input box, the increment is by one for each click, e.g. "10->11->12". Shouldn't this increase by the power of two, too?

LenzGr commented:
One more thing: the pool creation dialogue allows me to create pools without defining any application name, which may cause a health warning. Adding at least one application name should be mandatory.

Also, please replace "application metadata" with "applications / application names" to be in sync [with the documentation](http://docs.ceph.com/docs/master/rados/operations/pools/#associate-pool-to-application)

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.spec.ts:
```diff
@@ -0,0 +1,940 @@
+      }
+    ];
+    component.info['crush_rules_' + type].push(rule);
+    return rule;
```
Ty for spotting this. It was used before but the use has changed so this return is not longer needed.

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts:
```diff
@@ -0,0 +1,524 @@
+    }
+  }
+
+  _createForms() {
```
I can make it private and change the tests to test `component['createForms']()`

Devp00l commented:
@votdev You have to reload the backend because of a task name change.

Devp00l commented:
Addressed all comments :)

LenzGr commented:
Tested the latest version of that branch on my local dev system again today, looks good! I'd be in favor of merging this PR soon and then address any additional improvements and issues in separate PRs.

However, I noticed one issue: after creating a pool, the message in the "background tasks" list showed "Created pool 'undefined'" instead of the chosen pool name.

![screenshot from 2018-09-17 15-38-02](https://user-images.githubusercontent.com/263427/45626491-ffd46780-ba8f-11e8-9ef2-8e526a07febb.png)

LenzGr commented:
> However, I noticed one issue: after creating a pool, the message in the "background tasks" list showed "Created pool 'undefined'" instead of the chosen pool name.

Also, there is a slight mismatch of how the time is displayed and that the pool name in the recent notifications list is encapsulated by single quotes instead of double quotes:

![screenshot from 2018-09-17 15-42-13](https://user-images.githubusercontent.com/263427/45626765-956ff700-ba90-11e8-8429-a99de726af21.png)

LenzGr commented:
> When i edit a pool i never get the `Updated Pool 'xxx'` notification. The last notification is `Updating Pool 'xxx'`.

@votdev - it was my understanding that we don't want to generate notifications about successful events, in order to minimize the noise. Probably something to discuss and write down in the style guide?

Devp00l commented:
@LenzGr as mentioned in my [comment](https://github.com/ceph/ceph/pull/21614#issuecomment-420708092) you have to reload the backend first. This should solve all your problems.

The single quotes are also used for the RBDs, that means the user notification uses the wrong quoting.

ricardoasmarques commented:
When I tried to edit the following pool I got an error from an **'unknown task'**.

![screenshot from 2018-09-18 10-09-26](https://user-images.githubusercontent.com/14297426/45677066-37dcb880-bb2b-11e8-90f4-fbe95d1bba2a.png)

Shouldn't this error message display a proper task name?

Devp00l commented:
@ricardoasmarques I updated the PR description, that you have to reload the backend first, as you are the third tester that failed for it ;)

Devp00l commented:
@LenzGr IMHO it makes sense to create a success message for long running tasks

ricardoasmarques commented:
@Devp00l PR https://github.com/ceph/ceph/pull/23779 has been merged, please rebase this PR accordingly, thanks.

votdev commented:
> > When i edit a pool i never get the `Updated Pool 'xxx'` notification. The last notification is `Updating Pool 'xxx'`.
> 
> @votdev - it was my understanding that we don't want to generate notifications about successful events, in order to minimize the noise. Probably something to discuss and write down in the style guide?

I agree that this needs to be discussed.

Devp00l commented:
@ricardoasmarques rebased it and used the new deletion Dialog

p-na commented:
lgtm, except for the tooltip of the Crush Ruleset. That one is IMO confusing due to the text displayed and would need to get a proper border for top and left, too.

![screen](https://user-images.githubusercontent.com/9606095/45814343-a2315c80-bcd5-11e8-8bf8-e4ad8c7cd04f.png)

p-na commented:
One more nit: When creating an application tag for a pool the text says "Create application**s**" but it's only possible to create one application tag.

p-na commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,485 @@
+              </select>
+              <span class="help-block"
+                    i18n
+                    *ngIf="form.showError('crushRule', formDir, 'toFewOsds')">
```
`tooFewOsds`

Devp00l commented:
Addressed all comments

tspmelo commented:
jenkins retest this please

callithea commented:
jenkins test dashboard

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts:
```diff
@@ -0,0 +1,530 @@
+    ]);
+  }
+
+  private compareBlobSize(minimum, maximum) {
```
Better use this code because maximum may contain the string '-' which results in a comparision of 123456 >= null.

```
  private compareBlobSize(minimum, maximum) {
    minimum = this.formatter.toBytes(minimum);
    maximum = this.formatter.toBytes(maximum);
    if (_.isNull(minimum) || _.isNull(maximum)) {
      return false;
    }
    return minimum >= maximum;
  }
```

callithea commented:
jenkins test dashboard

callithea commented:
jenkins test dashboard

Devp00l commented:
Addressed all comments :+1:

p-na commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -0,0 +1,477 @@
+            </label>
+            <div class="col-sm-9">
+              <!-- <div class="input-group"> -->
+              <div>
```
Comment (L233) and container without attributes (L234) may not be required here.

callithea commented:
@Devp00l I removed the QA labels again due to the make check failure (``ERROR:   py27-lint: commands failed``,`` run-tox-mgr-dashboard (Failed)``).

Devp00l commented:
Addressed all comments

LenzGr commented:
`make check` failed:
```
ERROR: src/app/ceph/pool/pool-form/pool-form.component.html[347]: Attributes should vertically align
ERROR: src/app/ceph/pool/pool-form/pool-form.component.html[445]: Attributes should vertically align
ERROR: src/app/ceph/pool/pool-list/pool-list.component.html[11]: Attributes should vertically align
ERROR: src/app/ceph/pool/pool-list/pool-list.component.html[12]: Attributes should vertically align
ERROR: src/app/ceph/pool/pool-list/pool-list.component.html[13]: Attributes should vertically align
```

Devp00l commented:
Amended html so it passes the new html lint requirement.

s0nea commented:
Only one of four tests failed last run. I'm not really sure if it's related to the PR or if I just give it another try. What do you think?

- https://shaman.ceph.com/builds/ceph/wip-tdehler-testing-21614/29fbce64b836e9428cb826d25c6f572407195985/
- http://pulpito.ceph.com/tdehler-2018-10-05_06:10:43-rados:mgr-wip-tdehler-testing-21614-distro-basic-smithi/

Devp00l commented:
Hmm... The QA system we use are not the fastest ... but a internal timeout will kill the pool update test which has a timeout on it's own. @callithea can you help exceeding this internal timeout?

> 2018-10-05T12:22:43.619 INFO:teuthology.orchestra.run.smithi197:Running: "sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph log 'Ended test tasks.mgr.dashboard.test_pool.PoolTest.test_update'"

The only other solution is to comment out the `pg_num` update again.

callithea commented:
jenkins test dashboard

s0nea commented:
Sorry, but the tests are still failing :/

- https://shaman.ceph.com/builds/ceph/wip-tdehler-testing-21614/6f5a03957d6a24be9d674a7452b03944a1adaf5a/
- http://pulpito.ceph.com/tdehler-2018-10-08_11:20:33-rados:mgr-wip-tdehler-testing-21614-distro-basic-smithi/

callithea commented:
jenkins test dashboard
